platform/import-export/src/actions/ExportAsJSONAction.js

```diff
@@ -0,0 +1,170 @@
+ * licenses. See the Open Source Licenses file (LICENSES.md) included with
+ * this source code distribution or the Licensing information page available
+ * at runtime from the About dialog for additional information.
+ *****************************************************************************/
+
+define([], function () {
+
+    /**
+     * The ExportAsJSONAction is available from context menus and allows a user
+     * to export any creatable domain object as a JSON file.
+     *
+     * @implements {Action}
+     * @constructor
+     * @memberof platform/import-export
+     */
+    function ExportAsJSONAction(exportService, policyService, identifierService, context) {
+        this.root = {};
+        this.calls = 0;
+        this.context = context;
+        this.externalIdentifiers = [];
+        this.exportService = exportService;
+        this.policyService = policyService;
+        this.identifierService = identifierService;
+    }
+
+    ExportAsJSONAction.prototype.perform = function () {
+        this.contructJSON(this.context.domainObject);
+    };
+
+    ExportAsJSONAction.prototype.contructJSON = function (rootObject) {
+        var tree = {};
+        tree[rootObject.getId()] = rootObject.getModel();
+        this.root = rootObject;
+        // Root is included in tree during building to form relationships,
+        // removed after tree is built and re-added with "root" wrapper
+
+        this.write(tree, rootObject, [], function (result) {
+            this.exportService.exportJSON(result,
+                {filename: rootObject.getModel().name + '.json'});
+        }.bind(this));
+    };
+
+    /**
+     * Traverses the object hierarchy and populates the tree object with models
+     * and identifiers.
+     *
+     * @private
+     * @param {Object} tree
+     * @param {Object} parent
+     * @param {string []} seen
+     * @param callback
+     */
+    ExportAsJSONAction.prototype.write = function (tree, parent, seen, callback) {
+
+        this.calls++;
+        if (parent.hasCapability('composition')) {
+            parent.useCapability('composition')
+                .then(function (children) {
+                    children.forEach(function (child, index) {
+                        // Only export if object is creatable
+                        if (this.isCreatable(child)) {
+                            // If object is a link to something absent from
+                            // tree, generate new id and treat as new object
+                            if (this.isExternal(child, parent, tree)) {
+                                this.rewriteLink(child, parent, tree);
+                            } else {
+                                tree[child.getId()] = child.getModel();
+                            }
+                            // Prevents infinite export of objects that contain
+                            // themselves
+                            seen.push(parent.getId());
```

Comment:
I'd flatten out write.  You can store `tree`, and `callback` as properties on `this` (in perform).  
`seen` is unnecessary, you can check for `tree.hasOwnProperty(id)`.

Then `write` becomes `write(object)`.

After you refactor that, then it would be trivial to turn this into a promise chain, but I can talk you through that.